src/Akka.Streams.Kafka/Stages/Consumers/CommitCollectorStage.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using Akka.Streams.Kafka.Messages;
+using Akka.Streams.Kafka.Settings;
+using Akka.Streams.Stage;
+
+namespace Akka.Streams.Kafka.Stages.Consumers;
+
+/// <summary>
+/// INTERNAL API
+///
+/// Combined stage for committing incoming offsets in batches.
+/// </summary>
+internal sealed class CommitCollectorStage : GraphStage<FlowShape<ICommittable, ICommittableOffsetBatch>>
+{
+    private readonly CommitterSettings _settings;
+
+    public Inlet<ICommittable> In { get; } = new("FlowIn");
+    public Outlet<ICommittableOffsetBatch> Out { get; } = new("FlowOut");
+
+    public CommitCollectorStage(CommitterSettings settings)
+    {
+        _settings = settings;
+        Shape = new FlowShape<ICommittable, ICommittableOffsetBatch>(In, Out);
+    }
+
+    public override FlowShape<ICommittable, ICommittableOffsetBatch> Shape { get; }
+
+    protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) =>
+        new CommitCollectorStageLogic(Shape, this, inheritedAttributes);
+
+    private class CommitCollectorStageLogic : TimerGraphStageLogic
+    {
+        private const string CommitNow = "FlowStageCommit";
+
+        public CommitCollectorStageLogic(Shape shape, CommitCollectorStage stage, Attributes inheritedAttributes) :
+            base(shape)
+        {
+            Stage = stage;
+            InheritedAttributes = inheritedAttributes;
+            ObservationLogic = new CommitObservationLogic(stage._settings);
+            SetHandler(Stage.Out, OnPull);
+            SetHandler(Stage.In, OnPush, OnUpstreamFinish, OnUpstreamFailure);
+        }
+
+        private void OnUpstreamFailure(Exception ex)
+        {
+            Log.Debug(ex, "OnUpstreamFailure with exception with {0}", ObservationLogic.OffsetBatch);
+            if (ActiveBatchInProgress)
```

Comment:
If we have active data and the stages above us fail, deliver our pending batch in the form of an emergency commit to the `KafkaConsumerActor`. We have tests that verify that this works.